### PR TITLE
Enhance split_string to support splitting on white space [TG-2922]

### PR DESCRIPTION
**Pull request by thk123**
Split_string now won't error if you give it a whitespace delimiter and don't require it strip whitespace. Migrated the unit tests into Catch and added a few more examples. Made the utility throw exceptions rather than immediately terminate to allow the error cases to be unit tested as well. 

The unit tests document the _current_ behaviour. However, in some edge cases I believe the current behaviour is confusing. If someone knows for certain we don't depend on that behaviour I can prepare a PR to "fix" the behaviour. 

Review commit by commit.

### CODE REVIEW

**Review by majakusber — APPROVED** [review]
Looks good, I like the thorough tests. Consider squashing the commits to have only one function adjustment commit and one unit test commit.

**Comment by majakusber** [comment]
Nit: The `string_utils.cpp` file has Daniel Poetzl as author, should that be changed to Diffblue Ltd.?

**Comment by thk123** [comment]
@svorenova No - it was (I assume) written by @danpoe before he started at Diffblue so I assume can be left unchanged.

**Comment by thk123** [comment]
@svorenova Also respectfully disagree with squashing suggestion - I think current structure allows for checking a) correctly migrate existing unit tests b) correctly refactor existing unit tests c) correctly add unit tests for current behaviour d) correctly modify behaviour

Instead I propose squashing into:
- Removing wrong elements from the make file (incidental fix so will keep separate). 
- migrate unit tests
- refactor unit tests
- extend unit tests
- modify function directly to throwing exceptions and adding comments and unit tests for new behaviour

LMK if that's OK

**Comment by majakusber** [comment]
Yes, that sounds great, thank you.

**Comment by thk123** [comment]
Rebased as suggested.

File: unit/util/string_utils/split_string.cpp
Comment (danpoe) [inline unit/util/string_utils/split_string.cpp:114]: I agree, we could change this to return an empty vector. I've looked at the places where `split_string()` is currently used, and they would also work when returning an empty vector (except an assertion `assert(!result.empty())` would need to be removed in `util/unwind.cpp`). Also once changed, the current usages of `split_string()` can be slightly simplified as some of them have a check for this case.

**Review by danpoe — APPROVED** [review]
Looks good to me!

**Comment by thk123** [comment]
Specifically requesting a review from @chrisr-diffblue (somewhat arbitrarily).

File: src/util/string_utils.cpp
Comment (tautschnig) [inline src/util/string_utils.cpp]: I would have used `PRECONDITION` here as we don't generally throw exceptions unless it's triggerable by user input?

File: unit/util/string_utils/split_string.cpp
Comment (thk123) [reply unit/util/string_utils/split_string.cpp:114]: Excellent. Since this PR is already a pre-req for another I will put together a separate PR to address this.

File: src/util/string_utils.cpp
Comment (thk123) [reply src/util/string_utils.cpp]: Yeah reasonable - of course the method should just return the vector - if not hard will make as part of subsequent PR for empty elements.

File: src/util/string_utils.cpp
Comment (chrisr-diffblue) [inline src/util/string_utils.cpp]: Is there a reason that this condition/throw block can't just be an INVARIANT ?

File: src/util/string_utils.cpp
Comment (chrisr-diffblue) [inline src/util/string_utils.cpp:103]: Cheeky of me, but as you are touching this function, do you fancy adding some Doxygen? ;-)

**Review by chrisr-diffblue — APPROVED** [review]
A couple of minor comments, but nothing that would block this from my point of view.

File: src/util/string_utils.cpp
Comment (thk123) [reply src/util/string_utils.cpp]: For the unit test

**Comment by thk123** [comment]
Gah got the code ownership wrong, still need someone to vouch for unit changes. @tautschnig, as you've already looked at this a bit would you mind?

File: src/util/string_utils.cpp
Comment (thk123) [reply src/util/string_utils.cpp:103]: Will pick this up as part of the follow on tidy up PR as this is blocking another PR #2046

File: src/util/string_utils.cpp
Comment (chrisr-diffblue) [reply src/util/string_utils.cpp]: Hmm - ok, though that smells a little bit of the mechanics of unit testing driving the architecture of the code, but I don't have a hugely strong opinion. I just felt that an INVARIANT, or perhaps actually PRECONDITION now I think more, was more inline with other functions in this file that do things like `PRECONDITION(result.empty());`

Comment (tautschnig) [reply src/util/string_utils.cpp]: I am with @chrisr-diffblue: I think this should be a `PRECONDITION`. I think `invariant.h` ought to have the facilities to turn that into something that can be caught, and hence could be used for testing error paths in unit tests.

Comment (thk123) [reply src/util/string_utils.cpp]: I don't agree - I think of all the code we write, `util` code is the most "library like" and not throwing limits the utility of this function. Imagine Using `split_string` on a user input: I now have to duplicate the error handling code if I want to allow the user to try again.

Comment (chrisr-diffblue) [reply src/util/string_utils.cpp]: I think @thk123 does make a good point about the 'library like' use-case - perhaps the answer would be to have 'CATCHABLE_PRECONDITION' along with @tautschnig suggestion to enable error path testing? - i.e. in the library use-case Thomas is talking about, it could be caught by the caller to avoid duplicating the error handling.

Comment (tautschnig) [reply src/util/string_utils.cpp]: There are two different "users" to be distinguished: the developer using the library and the end-user feeding input to what the developer has built. The end-user should not be subjected to internal limitations, but for a developer it's reasonable to assume that they have read the documentation.

It is of course perfectly reasonable to use exceptions for either class of "users," but that's not what the code base does right now. If you think this should change, I'd go via a new issue and/or have some further discussion. Because your proposed code is of course very nice and clean I'd keep it so that a future patch to introduce the exceptions is very low effort.

Comment (tautschnig) [reply src/util/string_utils.cpp]: In all this I do think that `PRECONDITION_STRUCTURED` should be able to support this. There are a few cases of `INVARIANT_STRUCTURED` in the code base. (But also `invariant.h` says something about the future...)

Comment (thk123) [reply src/util/string_utils.cpp]: Sorry I'm not clear on what you propose I do with this code for this PR? Do you mean keep the code in this PR, or change this PR to a precondition (and remove the associated unit tests) and stash this version away for a later day?

Comment (tautschnig) [reply src/util/string_utils.cpp]: Here is what I'd do if it were my PR: if things can be made to work via `PRECONDITION_STRUCTURED`, use it. Else stash it away for a later day and drop a note in an issue so that it gets tracked.

**Comment by thk123** [comment]
@tautschnig Unfortunately `PRECONDITION_STRUCTURED` also `abort()`s and doesn't seem to be an existing version of the invariant logic that actually `throw`s the exception. (Also, I'm required to make the exception derive from `logic_error` which is clearly incorrect for invalid arguments. In the interest of getting this PR moving I've changed it to a PRECONDITION. I will create a follow on PR where we can discuss a better way of doing this.

**Comment by thk123** [comment]
@tautschnig If you're happy (as implied by a :+1:) could you approve? Apparently don't yet have sufficient code owner approval.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**
